nxc/protocols/smb/kerberos.py

```diff
@@ -0,0 +1,381 @@
+import datetime
+import logging
+import struct
+import random
+from impacket.winregistry import hexdump
+from six import b
+
+from pyasn1.codec.der import decoder, encoder
+from pyasn1.type.univ import noValue
+
+from impacket.krb5.asn1 import AP_REQ, AS_REP, TGS_REQ, Authenticator, TGS_REP, seq_set, seq_set_iter, PA_FOR_USER_ENC, \
+    Ticket as TicketAsn1, EncTGSRepPart, PA_PAC_OPTIONS
+from impacket.krb5.types import Principal, KerberosTime, Ticket
+from impacket.krb5.kerberosv5 import sendReceive, getKerberosTGT
+from impacket.krb5.ccache import CCache
+from impacket.krb5.crypto import Key, _enctype_table, _HMACMD5
+from impacket.krb5 import constants
+
+def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash, aesKey, kdcHost, impersonate, spn, useCache, no_s4u2proxy = False):
+    TGT = None
+    if useCache:
+        domain, user, tgt, _ = CCache.parseFile(domain, user, 'cifs/%s' % hostname)
+        if TGT is None:
+            raise
+        TGT = tgt['KDC_REP']
+        cipher = tgt['cipher']
+        sessionKey = tgt['sessionKey']
+    if TGT is None:
+        userName = Principal(username, type=constants.PrincipalNameType.NT_PRINCIPAL.value)
+        logging.info('Getting TGT for user')
+        tgt, cipher, _, sessionKey = getKerberosTGT(userName, password, domain,
+                                                                lmhash, nthash,
+                                                                aesKey,
+                                                                kdcHost)
+        TGT = decoder.decode(tgt, asn1Spec=AS_REP())[0]
+    decodedTGT=TGT
+    # Extract the ticket from the TGT
+    ticket = Ticket()
+    ticket.from_asn1(decodedTGT['ticket'])
+
+    apReq = AP_REQ()
+    apReq['pvno'] = 5
+    apReq['msg-type'] = int(constants.ApplicationTagNumbers.AP_REQ.value)
+
+    opts = list()
+    apReq['ap-options'] = constants.encodeFlags(opts)
+    seq_set(apReq, 'ticket', ticket.to_asn1)
+
+    authenticator = Authenticator()
+    authenticator['authenticator-vno'] = 5
+    authenticator['crealm'] = str(decodedTGT['crealm'])
+
+    clientName = Principal()
+    clientName.from_asn1(decodedTGT, 'crealm', 'cname')
+
+    seq_set(authenticator, 'cname', clientName.components_to_asn1)
+
+    now = datetime.datetime.utcnow()
+    authenticator['cusec'] = now.microsecond
+    authenticator['ctime'] = KerberosTime.to_asn1(now)
+
+    if logging.getLogger().level == logging.DEBUG:
+        logging.debug('AUTHENTICATOR')
+        print(authenticator.prettyPrint())
+        print('\n')
```

Comment:
Same for this, or is there a reason we would need another logger?

Reply:
resolved